### PR TITLE
Limit CanvasContext to max browser width/height, warn

### LINKED ISSUE #742: Canvas Backend breaks at height > 32767 (browser limit)
Using the Canvas backend, you can resize a canvas to height=32767 and it renders fine, but resize to height=32768 and it breaks, showing this in Chrome:
image

Try it here:
https://jsfiddle.net/2Lsme8yd/
(setting the height back to 32767 fixes the problem)

This is because browsers limit the size of a canvas element:
https://stackoverflow.com/a/11585939/10295942

(It's funny that they just break without a real error message though.
And that a 16 bit signed integer seems to be the somewhat arbitrary limit. I guess you can fit width+height in one 32-bit register then.)

Btw, there's a npm module canvas-size that tells you the current browser's maximum canvas size, which could be used:
https://stackoverflow.com/a/53677532/10295942

So, @0xfe what do you think would be the best way to deal with this? OSMD logs a warning and simply reduces the height (and width) to the maximum of 32767.
I suppose Vexflow could do the same, outputting a warning with Vex.W(). Or exit with 1? ;)

I can make a PR for this, just wanted to discuss how to proceed first.

### PR DESCRIPTION
**Pull request by sschmidTU**
fix #742

old behavior:
browser canvas element crashes when width or height > 32767 used on renderer.resize()
https://jsfiddle.net/2Lsme8yd/

new behavior:
renderer.resize() automatically crops dimensions to 32767 and gives warning
https://jsfiddle.net/tm9qaLzb/

npm test ran without fails.

(note that this only applies to CanvasContext, not SVGContext)

### CODE REVIEW

File: src/canvascontext.js [hunk 1]
```diff
@@ -13,6 +15,26 @@ export class CanvasContext {
...
     return 400;
   }
+  static get CANVAS_BROWSER_SIZE_LIMIT() {
+    return 32767; // Chrome/Firefox. Could be determined more precisely by npm module canvas-size
+  }
+
+  static CheckCanvasDimensionsForBrowserLimit(width, height) {
+    if (Math.max(width, height) > this.CANVAS_BROWSER_SIZE_LIMIT) {
+      Vex.W(
+        'canvas dimensions exceed browser limit. reducing/cropping to ' +
```
Comment (0xfe) [inline src/canvascontext.js]: Fix capitalization please 'Canvas dimensions exceed browser limit. Cropping to ...'

File: src/canvascontext.js [hunk 2]
```diff
@@ -13,6 +15,26 @@ export class CanvasContext {
   static get HEIGHT() {
     return 400;
   }
+  static get CANVAS_BROWSER_SIZE_LIMIT() {
+    return 32767; // Chrome/Firefox. Could be determined more precisely by npm module canvas-size
+  }
+
+  static CheckCanvasDimensionsForBrowserLimit(width, height) {
```
Comment (0xfe) [inline src/canvascontext.js]: Maybe a shorter function name: `SanitizeCanvasDims`

File: src/canvascontext.js [hunk 3]
```diff
@@ -98,7 +120,10 @@ export class CanvasContext {
   }
 
   resize(width, height) {
-    return this.vexFlowCanvasContext.resize(parseInt(width, 10), parseInt(height, 10));
+    width = parseInt(width, 10);
+    height = parseInt(height, 10);
+    [width, height] = this.CheckCanvasDimensionsForBrowserLimit(width, height);
```
Comment (0xfe) [inline src/canvascontext.js]: You can move the `parseInt` into the above `SanitizeCanvasDims` (which is a better place for it.)

File: src/renderer.js [hunk 4]
```diff
@@ -45,7 +45,12 @@ export class Renderer {
 
   static buildContext(elementId, backend, width, height, background) {
     const renderer = new Renderer(elementId, backend);
-    if (width && height) { renderer.resize(width, height); }
+    if (width && height) {
+      if (backend === Renderer.Backends.CANVAS) {
+        [width, height] = CanvasContext.CheckCanvasDimensionsForBrowserLimit(width, height);
```
Comment (0xfe) [inline src/renderer.js]: Looks like you're already calling this inside `resize`, so we don't need this here, no?

Comment (sschmidTU) [reply src/renderer.js]: ah, right, i thought this was separate. removing this call.

File: src/canvascontext.js (hunk 3 above)
Comment (sschmidTU) [reply src/canvascontext.js]: I assume like this:
`[width, height] = this.SanitizeCanvasDims(parseInt(width, 10), parseInt(height, 10));`
becuse moving it into the `SanitizeCanvasDims` function would mean i 'd need to call parseInt again for width and height for `this.vexFlowCanvasContext.resize(width, height);`, which would be duplicated.

File: src/canvascontext.js (hunk 2 above)
Comment (sschmidTU) [reply src/canvascontext.js]: yup, sounds good

File: src/canvascontext.js (hunk 1 above)
Comment (sschmidTU) [reply src/canvascontext.js]: done

**Comment by sschmidTU** [comment]
changes are all implemented, and tested again.
thanks for reviewing by the way!

File: src/canvascontext.js (hunk 3 above)
Comment (0xfe) [reply src/canvascontext.js]: Okay, that's fine, thanks.

**Comment by 0xfe** [comment]
Looks good. Thanks for the PR! Merging :-)

**Pull request merged**